### PR TITLE
Added Firmware upgrade test to nvme test

### PR DESCRIPTION
**Pull request by narasimhan-v**
* Added firmware upgrade test to nvme test.
* Added controller reset via sysfs test to nvme test.

Signed-off-by: Narasimhan V <sim@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by narasimhan-v** [comment]
```
2017-06-11 00:28:59,664 process          L0389 INFO | Running 'nvme list | grep /dev/nvme0'
2017-06-11 00:28:59,668 process          L0479 DEBUG| [stdout] /dev/nvme0n1     STM0001A8A0F         PCIe3 1.6TB NVMe Adapter                 1           1.60  TB /   1.60  TB      4 KiB +  0 B   KMIPP113
2017-06-11 00:28:59,669 process          L0499 INFO | Command 'nvme list | grep /dev/nvme0' finished with 0 after 0.00265502929688s
2017-06-11 00:28:59,669 nvmetest         L0097 DEBUG| Current FW: KMIPP113
2017-06-11 00:28:59,669 process          L0389 INFO | Running 'nvme fw-log /dev/nvme0'
2017-06-11 00:28:59,673 process          L0479 DEBUG| [stdout] Firmware Log for device:nvme0
2017-06-11 00:28:59,673 process          L0499 INFO | Command 'nvme fw-log /dev/nvme0' finished with 0 after 0.00178408622742s
2017-06-11 00:28:59,674 process          L0479 DEBUG| [stdout] afi  : 0x22
2017-06-11 00:28:59,674 process          L0479 DEBUG| [stdout] frs1 : 0x303031504e474d4b (KMGNP100)
2017-06-11 00:28:59,674 process          L0479 DEBUG| [stdout] frs2 : 0x3331315050494d4b (KMIPP113)
2017-06-11 00:28:59,674 process          L0479 DEBUG| [stdout] frs3 : 0x3330315050494d4b (KMIPP103)
2017-06-11 00:28:59,674 process          L0479 DEBUG| [stdout] frs4 : 0x3730315050494d4b (KMIPP107)
2017-06-11 00:28:59,675 process          L0389 INFO | Running 'nvme fw-download /dev/nvme0 --fw=/var/tmp/avocado_PQEGt1/KMIPP111.bin'
2017-06-11 00:28:59,690 process          L0479 DEBUG| [stdout] Firmware download success
2017-06-11 00:28:59,690 process          L0499 INFO | Command 'nvme fw-download /dev/nvme0 --fw=/var/tmp/avocado_PQEGt1/KMIPP111.bin' finished with 0 after 0.0135319232941s
2017-06-11 00:28:59,691 process          L0389 INFO | Running 'nvme fw-activate /dev/nvme0 -a 1 -s 2'
2017-06-11 00:29:06,986 process          L0479 DEBUG| [stdout] Success activating firmware action:1 slot:2
2017-06-11 00:29:06,986 process          L0499 INFO | Command 'nvme fw-activate /dev/nvme0 -a 1 -s 2' finished with 0 after 7.29255318642s
2017-06-11 00:29:06,986 process          L0389 INFO | Running 'echo 1 > /sys/class/nvme/nvme0/reset_controller'
2017-06-11 00:29:11,358 process          L0499 INFO | Command 'echo 1 > /sys/class/nvme/nvme0/reset_controller' finished with 0 after 4.36967897415s
2017-06-11 00:29:11,359 process          L0389 INFO | Running 'nvme fw-log /dev/nvme0'
2017-06-11 00:29:11,365 process          L0479 DEBUG| [stdout] Firmware Log for device:nvme0
2017-06-11 00:29:11,365 process          L0499 INFO | Command 'nvme fw-log /dev/nvme0' finished with 0 after 0.00362801551819s
2017-06-11 00:29:11,365 process          L0479 DEBUG| [stdout] afi  : 0x22
2017-06-11 00:29:11,365 process          L0479 DEBUG| [stdout] frs1 : 0x303031504e474d4b (KMGNP100)
2017-06-11 00:29:11,365 process          L0479 DEBUG| [stdout] frs2 : 0x3131315050494d4b (KMIPP111)
2017-06-11 00:29:11,366 process          L0479 DEBUG| [stdout] frs3 : 0x3330315050494d4b (KMIPP103)
2017-06-11 00:29:11,366 process          L0479 DEBUG| [stdout] frs4 : 0x3730315050494d4b (KMIPP107)
2017-06-11 00:29:11,366 process          L0389 INFO | Running 'nvme list | grep /dev/nvme0'
2017-06-11 00:29:11,373 process          L0479 DEBUG| [stdout] /dev/nvme0n1     STM0001A8A0F         PCIe3 1.6TB NVMe Adapter                 1           1.60  TB /   1.60  TB      4 KiB +  0 B   KMIPP111
2017-06-11 00:29:11,373 process          L0499 INFO | Command 'nvme list | grep /dev/nvme0' finished with 0 after 0.00473308563232s
2017-06-11 00:29:11,374 process          L0389 INFO | Running 'nvme show-regs /dev/nvme0 | grep version | awk '{print $NF}''
2017-06-11 00:29:11,378 process          L0479 DEBUG| [stdout] 10100
2017-06-11 00:29:11,378 process          L0499 INFO | Command 'nvme show-regs /dev/nvme0 | grep version | awk '{print $NF}'' finished with 0 after 0.00222086906433s
2017-06-11 00:29:11,384 test             L0750 INFO | PASS 1-nvmetest.py:NVMeTest.test_firmware_upgrade;run-99fc
```

**Comment by narasimhan-v** [comment]
```
2017-06-11 00:29:11,789 process          L0389 INFO | Running 'echo 1 > /sys/class/nvme/nvme0/reset_controller'
2017-06-11 00:29:16,163 process          L0499 INFO | Command 'echo 1 > /sys/class/nvme/nvme0/reset_controller' finished with 0 after 4.37237095833s
2017-06-11 00:29:16,164 process          L0389 INFO | Running 'nvme show-regs /dev/nvme0 | grep version | awk '{print $NF}''
2017-06-11 00:29:16,168 process          L0479 DEBUG| [stdout] 10100
2017-06-11 00:29:16,168 process          L0499 INFO | Command 'nvme show-regs /dev/nvme0 | grep version | awk '{print $NF}'' finished with 0 after 0.0022759437561s
2017-06-11 00:29:16,174 test             L0750 INFO | PASS 2-nvmetest.py:NVMeTest.testreset_sysfs;run-99fc
```

File: io/disk/ssd/nvmetest.py [hunk 1]
```diff
@@ -54,6 +56,66 @@ def setUp(self):
         self.format_size = pow(2, int(self.format_size))
         cmd = "nvme id-ns %s | grep 'in use' | awk '{print $2}'" % self.id_ns
         self.lba = process.system_output(cmd, shell=True).strip('\n')
+        self.firmware_url = self.params.get('firmware_url', default='')
+        if 'firmware_upgrade' in str(self.name) and not self.firmware_url:
+            self.skip("firmware url not gien")
```
Comment (PraveenPenguin) [inline io/disk/ssd/nvmetest.py]: is firmware_upgrade a feature for test if yes i don't think we can skip the test.
As well if already system is updated with latest firmware so i think no need to update that , what you think ?

Comment (narasimhan-v) [inline io/disk/ssd/nvmetest.py]: I am skipping the test only if firmware is not given.

Comment (PraveenPenguin) [inline io/disk/ssd/nvmetest.py]: sure what about if it is already on latest level ?

Comment (narasimhan-v) [inline io/disk/ssd/nvmetest.py]: I think we need not worry about that. The scope of this test is:
* Ensure firmware update works properly
* Ensure that by the end of the test, the controller is in the firmware given by the user.

I can take your suggestion as an enhancement, and I will work on it in the next commit.

Comment (PraveenPenguin) [inline io/disk/ssd/nvmetest.py]: sure make sense

**Pull request merged**